docs/src/userguide/extension_types.rst

```diff
@@ -1001,6 +1001,63 @@ disable this auto-generation and can be used to support pickling of more
 complicated types.
 
 
+Pattern matching and extension types (Python 3.10+)
+===================================================
+
+Python 3.10 introduced the structural pattern matching feature
+(documented in `the Python documentation
+<https://docs.python.org/3/reference/compound_stmts.html#the-match-statement>`_,
+and `PEP 634 <https://peps.python.org/pep-0634>`_).
+where Python objects can be decomposed if they meet a particular 
+set of rules. For a type to be treated as a sequence in pattern 
+matching it must have the flag ``Py_TPFLAGS_SEQUENCE`` in its 
+internal "type flags", and similarly for a type to be treated 
+as a mapping it must have ``Py_TPFLAGS_MAPPING`` set.
+
+There's a number of ways for these flags to be set: they're 
+set for a number of builtin types.  They will be set on a
+type where one of the base classes has the flag (including
+the builtin types, and also the relevant abstract base
+class in ``abc.collections``).
+
+A third way to mark a Python class as a
+sequence or mapping would be to register the class with 
+either ``abc.collections.Sequence`` or ``abc.collections.Mapping``
+using their ``.register`` function.
+Unfortunately for Cython extension types this does not work because
+Cython extension types are immutable (usually, with certain C defines
+set this may not always be true).
+
+Therefore a Cython extension type can be decorated with
+``cython.collection_type("sequence")`` or 
+``cython.collection_type("mapping")`` to enable pattern matching to
+use the type as intended.  These decorators are only needed
+to make the ``abc.collections`` type registration work
+(we recommend you use them *in addition* to type registration
+although pattern matching will work with the decorator alone).
+If your Cython class would become a sequence or mapping
+through inheritance then it does not need the decorators
+although you may choose to provide them for clarity.
+
+As an example::
```

Comment:
A small improvement - can we move example to files? It will be executed by CI then.

Reply:
The two reasons why I didn't:
* laziness ;)
* The second half of the example doesn't currently work in Cython itself (because pattern matching isn't yet merged) so would have to be excluded from the file.

It probably is worth compiling the class definition though - it at least guards against silly typos

Reply:
> The second half of the example doesn't currently work in Cython itself (because pattern matching isn't yet merged) so would have to be excluded from the file.

I think leaving Cython part compiling is reasonable compromise. With python part there is also problem running it only on python >= 3.10